Mathlib/LinearAlgebra/RootSystem/Hom.lean

```diff
@@ -82,31 +102,450 @@ def Hom.comp {ι₁ M₁ N₁ ι₂ M₂ N₂ : Type*} [AddCommGroup M₁] [Modu
     simp
 
 @[simp]
-lemma Hom.id_comp {ι₂ M₂ N₂ : Type*}
+lemma id_comp {ι₂ M₂ N₂ : Type*}
     [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
     (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂) (f : Hom P Q) :
-    Hom.comp f (Hom.id P) = f := by
+    comp f (id P) = f := by
   ext x <;> simp
 
 @[simp]
-lemma Hom.comp_id {ι₂ M₂ N₂ : Type*}
+lemma comp_id {ι₂ M₂ N₂ : Type*}
     [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
     (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂) (f : Hom P Q) :
-    Hom.comp (Hom.id Q) f = f := by
+    comp (id Q) f = f := by
   ext x <;> simp
 
 @[simp]
-lemma Hom.comp_assoc {ι₁ M₁ N₁ ι₂ M₂ N₂ ι₃ M₃ N₃ : Type*} [AddCommGroup M₁] [Module R M₁]
+lemma comp_assoc {ι₁ M₁ N₁ ι₂ M₂ N₂ ι₃ M₃ N₃ : Type*} [AddCommGroup M₁] [Module R M₁]
     [AddCommGroup N₁] [Module R N₁] [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
     [AddCommGroup M₃] [Module R M₃] [AddCommGroup N₃] [Module R N₃] {P : RootPairing ι R M N}
     {P₁ : RootPairing ι₁ R M₁ N₁} {P₂ : RootPairing ι₂ R M₂ N₂} {P₃ : RootPairing ι₃ R M₃ N₃}
     (h : Hom P₂ P₃) (g : Hom P₁ P₂) (f : Hom P P₁) :
-    Hom.comp (Hom.comp h g) f = Hom.comp h (Hom.comp g f) := by
+    comp (comp h g) f = comp h (comp g f) := by
   ext <;> simp
 
-/-- The endomorphism of a root pairing given by a reflection. -/
+/-- The endomorphism monoid of a root pairing. -/
+instance (P : RootPairing ι R M N) : Monoid (Hom P P) where
+  mul := comp
+  mul_assoc := comp_assoc
+  one := id P
+  one_mul := id_comp P P
+  mul_one := comp_id P P
+
+@[simp]
+lemma weightMap_one (P : RootPairing ι R M N) :
+    weightMap (P := P) (Q := P) 1 = LinearMap.id (R := R) (M := M) :=
+  rfl
+
+@[simp]
+lemma coweightMap_one (P : RootPairing ι R M N) :
+    coweightMap (P := P) (Q := P) 1 = LinearMap.id (R := R) (M := N) :=
+  rfl
+
+@[simp]
+lemma indexEquiv_one (P : RootPairing ι R M N) :
+    indexEquiv (P := P) (Q := P) 1 = Equiv.refl ι :=
+  rfl
+
+@[simp]
+lemma weightMap_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    weightMap (x * y) = weightMap x ∘ₗ weightMap y :=
+  rfl
+
+@[simp]
+lemma coweightMap_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    coweightMap (x * y) = coweightMap y ∘ₗ coweightMap x :=
+  rfl
+
+@[simp]
+lemma indexEquiv_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    indexEquiv (x * y) = indexEquiv x ∘ indexEquiv y :=
+  rfl
+
+/-- The endomorphism monoid of a root pairing. -/
+abbrev _root_.RootPairing.End (P : RootPairing ι R M N) := Hom P P
+
+/-- The weight space representation of endomorphisms -/
+def weightHom (P : RootPairing ι R M N) : End P →* (M →ₗ[R] M) where
+  toFun g := Hom.weightMap (P := P) (Q := P) g
+  map_mul' g h := by ext; simp
+  map_one' := by ext; simp
+
+lemma weightHom_injective (P : RootPairing ι R M N) : Injective (weightHom P) := by
+  intro f g hfg
+  ext x
+  · exact congrFun (congrArg DFunLike.coe hfg) x
```

Comment:
Hardly matters but if things get this fiddly, there should be a lemma. Luckily there is here:
```suggestion
  · exact LinearMap.congr_fun hfg x
```